claragenomics/io/bedgraphio.py

```diff
@@ -18,34 +18,62 @@
 import pandas as pd
 
 
-def expand_interval(interval):
+def expand_interval(interval, score=True):
     """
-    Function to expand an interval to single-base resolution
+    Function to expand an interval to single-base resolution and add scores
     Args:
-        interval (list or list-like containing chrom, start, and end)
+        interval (dict or DataFrame containing chrom, start, end and optionally scores)
+        score (bool): add score to each base of interval
     Returns:
-        expanded: pandas dataframe containing a row for every base in the interval
+        expanded: pandas DataFrame containing a row for every base in the interval
     """
     expanded = pd.DataFrame(columns=['chrom', 'start'])
-    expanded['start'] = range(interval[1], interval[2])
-    expanded['chrom'] = interval[0]
+    expanded['start'] = range(interval['start'], interval['end'])
+    expanded['end'] = expanded['start'] + 1
+    expanded['chrom'] = interval['chrom']
+    # If necessary, assign a score to each base in the interval
+    if score:
+        expanded['score'] = interval['scores']
     return expanded
 
 
-def intervals_to_bg(intervals_df, scores):
+def contract_interval(expanded_df, positive=True):
     """
-    Function to combine intervals and scores in bedGraph format
+    Function to contract a dataframe containing genomic positions and scores into smaller intervals with equal score
     Args:
-        intervals_df: Pandas dataframe containing columns for chrom, start, end
-        scores: numeric scores (at single-base resolution) to be added to bedGraph
+        expanded_df: Pandas dataframe containing chrom, start, end, score at base resolution.
+        positive (bool): if True, only regions with score>0 are retained
     Returns:
-        bg: pandas dataframe containing expanded intervals and scores where score>0.
+        intervals_df: Pandas dataframe with same columns; bases with same score are combined into one line.
     """
+    # For each base, attach the score assigned to the previous base
+    expanded_df['prevscore'] = [-1] + list(expanded_df['score'])[:-1]    
+    # Select bases where score changes - or the last base
+    intervals_df = expanded_df[(expanded_df['score'] != expanded_df['prevscore']) | (expanded_df.index==len(expanded_df)-1)].copy()
```

Comment:
Are these scores always integers or can they be floating point number? If they're always integer, then feel free to resolve this comment :).

If they can be floating point numbers, then it's best to use some proximity comparison - like that mentioned in https://docs.python.org/3/whatsnew/3.5.html#pep-485-a-function-for-testing-approximate-equality

Reply:
ping @avantikalal

Reply:
They will be integers in all of our use cases right now. It is theoretically possible that in some future use case they may not be integers, but as I don't know what that would be I think we can merge this.

Reply:
In `postprocess.py` we either round or threshold the scores before applying this function, so small differences will be removed.

Reply:
Perfect, thanks @avantikalal !